cmd/juju/cloud/add_test.go

```diff
@@ -7,172 +7,256 @@ import (
+//       new-york:
+//         endpoint: http://newyork/1.0
+//   garage-maas:
+//     type: mass
+//     auth-types: [oauth]
+//     endpoint: http://garagemaas
+// `[1:]
+// 	err = ioutil.WriteFile(sourceFile, []byte(source), 0600)
+// 	c.Assert(err, jc.ErrorIsNil)
+// 	return sourceFile
+// }
+
 func (s *addSuite) TestAddBadFilename(c *gc.C) {
-	addCmd := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	badFileErr := errors.New("")
+	fake.Call("ParseCloudMetadataFile", "somefile.yaml").Returns(map[string]cloudfile.Cloud{}, badFileErr)
+
+	addCmd := cloud.NewAddCloudCommand(fake)
 	_, err := testing.RunCommand(c, addCmd, "cloud", "somefile.yaml")
-	c.Assert(err, gc.ErrorMatches, "open somefile.yaml: .*")
+	c.Check(err, gc.Equals, badFileErr)
 }
 
 func (s *addSuite) TestAddBadCloudName(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	addCmd := cloud.NewAddCloudCommand()
-	_, err := testing.RunCommand(c, addCmd, "cloud", sourceFile)
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "testFile").Returns(map[string]cloudfile.Cloud{}, nil)
+
+	addCmd := cloud.NewAddCloudCommand(fake)
+	_, err := testing.RunCommand(c, addCmd, "cloud", "testFile")
 	c.Assert(err, gc.ErrorMatches, `cloud "cloud" not found in file .*`)
 }
 
 func (s *addSuite) TestAddExisting(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(), "homestack", sourceFile)
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "fake.yaml").Returns(homestackMetadata(), nil)
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+
+	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(fake), "homestack", "fake.yaml")
 	c.Assert(err, gc.ErrorMatches, `"homestack" already exists; use --replace to replace this existing cloud`)
 }
 
 func (s *addSuite) TestAddExistingReplace(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(), "homestack", sourceFile, "--replace")
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "fake.yaml").Returns(homestackMetadata(), nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+	numCallsToWrite := fake.Call("WritePersonalCloudMetadata", homestackMetadata()).Returns(nil)
+
+	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(fake), "homestack", "fake.yaml", "--replace")
 	c.Assert(err, jc.ErrorIsNil)
-	data, err := ioutil.ReadFile(osenv.JujuXDGDataHomePath("clouds.yaml"))
-	c.Assert(string(data), gc.Equals, `
-clouds:
-  homestack:
-    type: openstack
-    auth-types: [userpass, access-key]
-    endpoint: http://homestack
-    regions:
-      london:
-        endpoint: http://london/1.0
-      new-york:
-        endpoint: http://newyork/1.0
-`[1:])
+
+	c.Check(numCallsToWrite(), gc.Equals, 1)
```

Comment:
So, in this case, what the test is actually testing is that the old data is replaced by the new data in Write.  But that is not at all clear from what this test is telling us.  The only thing it is explicitly says it is checking is that write was called once.  It doesn't have a line that says "check if what was written was correct".  The problem is that the test is not obviously correct.  You look at it and you think "hey, they're not even testing that what was written was correct".  You actualy are, since the test will fail if the written value is incorrect, but that has been obscured.  That makes it hard for people to understand what is being tested, and it makes it hard for people to come back and update the test later.

It also means that when the test fails, it'll give a really terrible failure message.  Instead of a very obvious 'expected to write "foo: bar", instead wrote "baz: bat"', you'll get "expected 1, got 0".  As a developer, it's going to be very unclear that the test is failing because it wrote the wrong thing, rather than thinking it just didn't write at all.